File: tools/pipewire_snapshot.py
"""Fold pw-dump's initial array and any final change batches by registry id."""
import json


def parse_dump(text):
    try:
        snapshot = json.loads(text)
    except json.JSONDecodeError:
        decoder = json.JSONDecoder()
        objects = {}
        remainder = text.lstrip()
        if not remainder:
            raise ValueError("Empty PipeWire dump")
        while remainder:
            batch, end = decoder.raw_decode(remainder)
            if not isinstance(batch, list):
                raise ValueError("Expected a PipeWire object array")
            for item in batch:
                if not isinstance(item, dict) or type(item.get("id")) is not int or not 0 <= item["id"] <= 0xffffffff:
                    raise ValueError("PipeWire update has no valid registry id")
                if ("info" in item and item["info"] is None) or ("props" in item and item["props"] is None):
                    objects.pop(item["id"], None)
                else:
                    objects[item["id"]] = item
            remainder = remainder[end:].lstrip()
        snapshot = list(objects.values())
    if not isinstance(snapshot, list):
        raise ValueError("Expected a PipeWire object array")
    return snapshot

File: src/OpenXLR.Core/Mixing/PipeWireSnapshot.cs
using System.Text.Json;

namespace OpenXLR.Core.Mixing;

/// <summary>
/// pw-dump may append change arrays while the graph is changing. Fold those
/// batches by registry id, replacing full objects and removing tombstones.
/// </summary>
internal static class PipeWireSnapshot
{
    internal static JsonDocument Parse(ReadOnlySpan<byte> json)
    {
        var reader = new Utf8JsonReader(json,
            new JsonReaderOptions { AllowMultipleValues = true });
        JsonDocument first = JsonDocument.ParseValue(ref reader);
        bool returnedFirst = false;
        try
        {
            if (first.RootElement.ValueKind != JsonValueKind.Array)
                throw new JsonException("Expected a PipeWire object array.");
            // Most snapshots contain one array. Keep that document without
            // cloning and serializing the entire graph on every sweep.
            if (!reader.Read())
            {
                returnedFirst = true;
                return first;
            }

            var objects = new Dictionary<uint, JsonElement>();
            Apply(first.RootElement, objects);
            do
            {
                using JsonDocument batch = JsonDocument.ParseValue(ref reader);
                Apply(batch.RootElement, objects);
            } while (reader.Read());
            return JsonDocument.Parse(JsonSerializer.SerializeToUtf8Bytes(objects.Values));
        }
        finally
        {
            if (!returnedFirst) first.Dispose();
        }
    }

    private static void Apply(JsonElement batch, Dictionary<uint, JsonElement> objects)
    {
        if (batch.ValueKind != JsonValueKind.Array)
            throw new JsonException("Expected a PipeWire object array.");
        foreach (JsonElement item in batch.EnumerateArray())
        {
            if (item.ValueKind != JsonValueKind.Object || !item.TryGetProperty("id", out JsonElement key)
                || key.ValueKind != JsonValueKind.Number || !key.TryGetUInt32(out uint id))
                throw new JsonException("PipeWire update has no valid registry id.");
            if (item.TryGetProperty("info", out JsonElement info) && info.ValueKind == JsonValueKind.Null
                || item.TryGetProperty("props", out JsonElement props) && props.ValueKind == JsonValueKind.Null)
                objects.Remove(id);
            else
                objects[id] = item.Clone();
        }
    }
}
